diff --git a/node_modules/@gmod/bbi/esm/wasm/inflate-wasm-inlined.js b/node_modules/@gmod/bbi/esm/wasm/inflate-wasm-inlined.js
index b2d8cfa..07f8d05 100644
--- a/node_modules/@gmod/bbi/esm/wasm/inflate-wasm-inlined.js
+++ b/node_modules/@gmod/bbi/esm/wasm/inflate-wasm-inlined.js
@@ -357,10 +357,17 @@ async function init() {
         return wrapper_wasm;
     if (!initPromise) {
         initPromise = (async () => {
-            const response = await fetch(inflate_wasm_bg_namespaceObject);
-            const bytes = await response.arrayBuffer();
+            const b64 = inflate_wasm_bg_namespaceObject.replace(
+                'data:application/wasm;base64,',
+                '',
+            );
+            const bin = atob(b64);
+            const bytes = new Uint8Array(bin.length);
+            for (let i = 0; i < bin.length; i++) {
+                bytes[i] = bin.charCodeAt(i);
+            }
             const { instance } = await WebAssembly.instantiate(bytes, {
-                './inflate_wasm_bg.js': src_wasm_inflate_wasm_bg_namespaceObject,
+            './inflate_wasm_bg.js': src_wasm_inflate_wasm_bg_namespaceObject,
             });
             wrapper_wasm = instance.exports;
             __wbg_set_wasm(wrapper_wasm);
