diff --git a/node_modules/@gmod/bgzf-filehandle/esm/wasm/bgzf-wasm-inlined.js b/node_modules/@gmod/bgzf-filehandle/esm/wasm/bgzf-wasm-inlined.js
index b8a4365..5227798 100644
--- a/node_modules/@gmod/bgzf-filehandle/esm/wasm/bgzf-wasm-inlined.js
+++ b/node_modules/@gmod/bgzf-filehandle/esm/wasm/bgzf-wasm-inlined.js
@@ -350,11 +350,19 @@ async function init() {
     if (!initPromise) {
         initPromise = (async () => {
             // wasmData is a data URL from asset/inline
-            const response = await fetch(bgzf_wasm_bg_namespaceObject);
-            const bytes = await response.arrayBuffer();
+            const b64 = bgzf_wasm_bg_namespaceObject.replace(
+                'data:application/wasm;base64,',
+                '',
+            );
+            const bin = atob(b64);
+            const bytes = new Uint8Array(bin.length);
+            for (let i = 0; i < bin.length; i++) {
+            bytes[i] = bin.charCodeAt(i);
+            }
             const { instance } = await WebAssembly.instantiate(bytes, {
-                './bgzf_wasm_bg.js': src_wasm_bgzf_wasm_bg_namespaceObject,
+            './bgzf_wasm_bg.js': src_wasm_bgzf_wasm_bg_namespaceObject,
             });
+
             wrapper_wasm = instance.exports;
             __wbg_set_wasm(wrapper_wasm);
             return wrapper_wasm;
